core/src/main/java/com/linecorp/armeria/server/healthcheck/ScheduledHealthChecker.java

```diff
@@ -0,0 +1,216 @@
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.Futures;
+
+import com.linecorp.armeria.common.util.AbstractListenable;
+
+import io.netty.util.concurrent.EventExecutor;
+
+/**
+ * A {@link ListenableHealthChecker} whose state can be set by the result of supplied CompletionStage.
+ * This class can be shared by multiple {@link HealthCheckService}.
+ */
+final class ScheduledHealthChecker extends AbstractListenable<HealthChecker>
+        implements ListenableHealthChecker {
+
+    private final Supplier<? extends CompletionStage<HealthCheckStatus>> healthChecker;
+    private final Duration fallbackTtl;
+    private final EventExecutor eventExecutor;
+    private final Consumer<HealthChecker> onHealthCheckerUpdate;
+    private final AtomicBoolean isHealthy;
+    private final AtomicInteger requestCount;
+    private final AtomicReference<ScheduledHealthCheckerImpl> impl;
+
+    ScheduledHealthChecker(Supplier<? extends CompletionStage<HealthCheckStatus>> healthChecker,
+                           Duration facllbackTtl, EventExecutor eventExecutor) {
+        this.healthChecker = healthChecker;
+        this.fallbackTtl = facllbackTtl;
+        this.eventExecutor = eventExecutor;
+
+        isHealthy = new AtomicBoolean();
+        requestCount = new AtomicInteger();
+        impl = new AtomicReference<>();
+        onHealthCheckerUpdate = latestValue -> {
+            isHealthy.set(latestValue.isHealthy());
+            notifyListeners(latestValue);
+        };
+    }
+
+    @Override
+    public boolean isHealthy() {
+        return isHealthy.get();
+    }
+
+    void startHealthChecker() {
+        if (requestCount.getAndIncrement() != 0) {
+            return;
+        }
+        final ScheduledHealthCheckerImpl newlyScheduled =
+                new ScheduledHealthCheckerImpl(healthChecker, fallbackTtl, eventExecutor);
+        // This spin avoid the impl is not null when this instance is shared by multiple servers.
+        // e.g. Server A start -> Server A stop(just finish decrementAndGet and get 0) ->
+        // Server B start(just finis getAndIncrement and get 0). And there is a window that impl is not set to
+        // null by Server A's stopHealthChecker. So Server B's startHealthChecker need to spin until Server
+        // A's stopHealthChecker complete.
+
+        // And this method only allow the first request can start the schedule by requestCount#getAndIncrement,
+        // if caller calling startHealthChecker/stopHealthChecker correctly(each pair should be called
+        // sequentially). So there should be no spin longer than expected.
```

Comment:
```suggestion
        // This spin prevents the following race condition, which occurs when this instance is shared
        // by more than one server instance:
        //
        // 1. Server A starts.
        // 2. Server A stops. decrementAndGet() returns 0, but it didn't clear `impl` yet.
        // 3. Server B starts; getAndIncrement() returns 0, but `impl` isn't `null` yet.
        //
        // There should be no unexpectedly long spin here, as long as a caller makes sure to call
        // `stopHealthChecker()` for each `startHealthChecker()`, because this method is guarded by
        // `requestCount.getAndIncrement()` to allow only the first request to start to schedule
        // a task.
```